R/join_keys.R

```diff
@@ -0,0 +1,591 @@
+
+  x_merged <- Reduce(
+    init = join_keys(),
+    x = x,
+    f = function(.x, .y) {
+      assert_compatible_keys2(.x, .y)
+      utils::modifyList(.x, .y, keep.null = FALSE)
+    }
+  )
+
+  utils::modifyList(join_keys_obj, x_merged, keep.null = FALSE)
+}
+
+#' @rdname join_keys
+#' @export
+#'
+#' @examples
+#'
+#' c(join_key("a", "b", "c"), join_keys(join_key("a", "d2", "c")))
+c.join_key_set <- function(...) {
+  c.join_keys(...)
+}
+
+#' The Names of an `join_keys` Object
+#' @inheritParams base::`names<-`
+#' @export
+`names<-.join_keys` <- function(x, value) {
+  new_x <- unclass(x)
+  parent_list <- parents(x)
+  # Update inner keys
+  for (old_name in setdiff(names(new_x), value)) {
+    old_entry <- new_x[[old_name]]
+    new_name <- value[names(new_x) == old_name]
+
+    # Change 2nd-tier first
+    for (sub_name in names(old_entry)) {
+      names(new_x[[sub_name]])[names(new_x[[sub_name]]) == old_name] <- new_name
+    }
+
+    # Change in first tier
+    names(new_x)[names(new_x) == old_name] <- new_name
+
+    # changing name in the parents
+    if (length(parent_list)) {
+      names(parent_list)[names(parent_list) == old_name] <- new_name
+      parent_list <- lapply(parent_list, function(.x) {
+        if (identical(.x, old_name)) {
+          new_name
+        } else {
+          .x
+        }
+      })
+      attr(new_x, "__parents__") <- parent_list
+    }
+  }
+
+  class(new_x) <- c("join_keys", "list")
+  new_x
+}
+
+#' @rdname join_keys
+#' @details
+#' Getter for `join_keys` that returns the relationship between pairs of datasets.
+#'
+#' @inheritParams base::`[[`
+#' @param keep_all_foreign_keys (`logical`) flag that keeps foreign keys and other
+#' datasets even if they are not a parent of the selected dataset.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Getter for join_keys ----
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]][["ds2"]] <- "some_col"
+#' jk[["ds1"]][["ds3"]] <- "some_col2"
+#'
+#' jk["ds1"]
+#' jk[1:2]
+#' jk[c("ds1", "ds2")]
+`[.join_keys` <- function(x, i, keep_all_foreign_keys = FALSE) {
+  if (missing(i)) {
+    return(x)
+  }
+
+  if (is.null(i)) {
+    return(new_join_keys()) # replicate base R
+  }
+
+  checkmate::assert(
+    combine = "or",
+    checkmate::check_integerish(i),
+    checkmate::check_logical(i),
+    checkmate::check_character(i)
+  )
+  checkmate::assert_logical(keep_all_foreign_keys, len = 1)
+
+  # Convert integer/logical index to named index
+  if (checkmate::test_integerish(i) || checkmate::test_logical(i)) {
+    i <- names(x)[i]
+  }
+
+  # When retrieving a relationship pair, it will also return the symmetric key
+  new_jk <- new_join_keys()
+  queue <- i
+  bin <- character(0)
+
+  # Need to iterate on a mutating queue if subset of a dataset will also
+  #  select its parent as that parent might have relationships with others
+  #  already selected.
+  while (length(queue) > 0) {
+    ix <- queue[1]
+    queue <- queue[-1]
+
+    if (ix %in% bin) {
+      next
+    }
```

Comment:
@averissimo could you give an example when this could happen?
```suggestion
    if (ix %in% bin) {
      next
    }
```

Reply:
```r
jk <- join_keys()
jk[c("d1", "d1")]
```

Reply:
Hmmm, how do you think? Should it be error (assert unique) or skipped?

Reply:
I think it's ok to accept and handle.

Reply:
Nice example @vedhav

This is in-place to double processing of a dataset in the queue. Repeated subsets for one, but also when parents are not a DAG (corrupt) and avoids an infinite loop.

Reply:
isn't `queue <- unique(i)` solves the problem? `unique` outside of the loop is easier to comprehend I think

Reply:
Sure. `&& !ix_parent %in% c(queue, bin)` a couple lines after provides the DAG protection.

Done in 3393eda7